File: update.py
import sqlite3 as lite
import os
import re

def parse_floats(input_string):
    input_string = input_string.replace(',', '.')

    float_pattern = r'\b\d+\.\d+\b'
    int_pattern = r'(?<!\d\.)\b\d+\b(?!\.\d)'
    
    # Find all floats and integers with their positions
    floats = [(float(m.group()), m.start()) for m in re.finditer(float_pattern, input_string)]
    ints = [(int(m.group()), m.start()) for m in re.finditer(int_pattern, input_string)]
    
    # Combine and sort them based on their positions
    numbers = floats + ints
    numbers.sort(key=lambda x: x[1])
    
    # Extract the sorted numbers
    sorted_numbers = [num[0] for num in numbers]
    
    return sorted_numbers

con = lite.connect('db.sqlite3')

for file in os.listdir("hrady"):
    hrad = {}
    hrad['fotky'] = []
    hrad['vybaveni'] = ""
    state = 0
    state_read = True
    with open("./hrady/" + file, "r") as f:
        for line in f:
            if line == "\n" and state_read:
                state += 1
                state_read = False

            else:
                state_read = True

                if state == 0:
                    hrad['name'] = line[:-1]

                elif state == 1:
                    hrad['type'] = int(line[:-1])

                elif state == 2:
                    hrad['popis'] = line[:-1]

                elif state == 3:
                    pattern = re.compile("^(([0-9]*[,])?[0-9]+) m x (([0-9]*[,])?[0-9]+) m")
                    if pattern.match(line):
                        floats = parse_floats(line)
                        hrad['x'] = floats[0]
                        hrad['y'] = floats[1]

                    else:
                        hrad['special'] = line[:-1]
                
                elif state == 4:
                    hrad['z'] = parse_floats(line)[0]

                elif state == 5:
                    hrad['druh'] = line[:-1]

                elif state == 6:
                    hrad['material'] = line[:-1]

                elif state == 7:
                    hrad['vybaveni'] += line

                elif state == 8:
                    hrad['kapacita'] = line[:-1]

                elif state == 9:
                    hrad['skladem'] = line[:-1]

                elif state == 10:
                    line = line[0] + line[2:]

                    floats = parse_floats(line)

                    hrad['cena'] = floats[0]
                    hrad['cena_hodina'] = floats[1]

                elif state == 11:
                    hrad['fotky'].append(line[:-1])
        # 
        # for i in hrad:
        #     print(f'{i} {hrad[i]}')

        with con:
            cur = con.cursor()    
            cur.execute("INSERT INTO castles(name, type, canHire) VALUES(?, ?, ?)", [hrad['name'], hrad['type'], True])
            id = cur.lastrowid
            if 'special' in hrad:
                cur.execute("INSERT INTO parametry (id, x, y, z, special) VALUES(?, ?, ?, ?, ?)", 
                            [id, None, None, hrad['z'], hrad['special']])
            else:
                cur.execute("INSERT INTO parametry (id, x, y, z, special) VALUES(?, ?, ?, ?, ?)", 
                            [id, hrad['x'], hrad['y'], hrad['z'], None])

            cur.execute("INSERT INTO popis (id, hlavni, druh, material, vybaveni, kapacita) VALUES (?, ?, ?, ?, ?, ?)",
                        [id, hrad['popis'], hrad['druh'], hrad['material'], hrad['vybaveni'], hrad['kapacita']])
            
            cur.execute("INSERT INTO popis_pronajem (id, skladem, cena, cena_hodina) VALUES (?, ?, ?, ?)",
                        [id, hrad['skladem'], hrad['cena'], hrad['cena_hodina']])
            
            for fotka in hrad['fotky']:
                cur.execute("INSERT INTO images (castle_id, path) VALUES (?, ?)", 
                            [id, fotka])


#     const result = await db.run(`
#       INSERT INTO castles (name, content)
#       VALUES (?, ?)
#     `, [castle.name, castle.content]);

#     // Get the last inserted row id (i.e., the castle id)
#     const castleId = result.lastID;

#     // Insert into parametry table
#     await db.run(`
#       INSERT INTO parametry (id, x, y, z, special)
#       VALUES (?, ?, ?, ?, ?)
#     `, [castleId, parametry.x, parametry.y, parametry.z, parametry.special]);

#     // Insert into popis table
#     await db.run(`
#       INSERT INTO popis (id, hlavni, pudorys, vyska, druh, material, vybaveni, kapacita)
#       VALUES (?, ?, ?, ?, ?, ?, ?, ?)
#     `, [castleId, popis.hlavni, popis.pudorys, popis.vyska, popis.druh, popis.material, popis.vybaveni, popis.kapacita]);

#     // Insert into popis_pronajem table
#     await db.run(`
#       INSERT INTO popis_pronajem (id, cena, cena_hodina)
#       VALUES (?, ?, ?)
#     `, [castleId, popis_pronajem.cena, popis_pronajem.cena_hodina]);
#   };

File: update.js
import sqlite3 from 'sqlite3';
import { open } from 'sqlite';


  const db = await open({
    filename: 'my-database.db',
    driver: sqlite3.Database
  });

  // Function to insert a castle with its relations
  const insertCastleWithRelations = async (castleData) => {
    const { castle, parametry, popis, popis_pronajem } = castleData;

    // Insert into castles table
    const result = await db.run(`
      INSERT INTO castles (name, content)
      VALUES (?, ?)
    `, [castle.name, castle.content]);

    // Get the last inserted row id (i.e., the castle id)
    const castleId = result.lastID;

    // Insert into parametry table
    await db.run(`
      INSERT INTO parametry (id, x, y, z, special)
      VALUES (?, ?, ?, ?, ?)
    `, [castleId, parametry.x, parametry.y, parametry.z, parametry.special]);

    // Insert into popis table
    await db.run(`
      INSERT INTO popis (id, hlavni, pudorys, vyska, druh, material, vybaveni, kapacita)
      VALUES (?, ?, ?, ?, ?, ?, ?, ?)
    `, [castleId, popis.hlavni, popis.pudorys, popis.vyska, popis.druh, popis.material, popis.vybaveni, popis.kapacita]);

    // Insert into popis_pronajem table
    await db.run(`
      INSERT INTO popis_pronajem (id, cena, cena_hodina)
      VALUES (?, ?, ?)
    `, [castleId, popis_pronajem.cena, popis_pronajem.cena_hodina]);
  };

  // Example data for a castle and its relations
  const castleData = {
    castle: { name: 'Castle Black', content: 'The stronghold of the Night\'s Watch' },
    parametry: { x: 123.45, y: 67.89, z: 10.11, special: 'Special Feature' },
    popis: {
      hlavni: 'Main Hall',
      pudorys: 'Rectangular',
      vyska: '50 meters',
      druh: 'Stone',
      material: 'Granite',
      vybaveni: 'Armory',
      kapacita: '100 people'
    },
    popis_pronajem: { cena: 1000, cena_hodina: 50 }
  };

  // Insert example castle with relations
  await insertCastleWithRelations(castleData);

  // Close the database
  await db.close();

// Example usage
const newCastleData = {
  name: 'Castle Example',
  type: 0, // Example type
  parametry: {
    x: 10.0,
    y: 20.0,
    z: 30.0,
    special: 'Special features here',
  },
  popis: {
    hlavni: 'Author',
    pudorys: 'Layout',
    vyska: 'Height',
    druh: 'Type',
    material: 'Material',
    vybaveni: 'Equipment',
    kapacita: 'Capacity',
  },
  popis_pronajem: {
    skladem: 'In stock',
    cena: 1000,
    cena_hodina: 100,
  },
  images: [
    { path: 'image1.jpg' },
    { path: 'image2.jpg' },
  ],
};

fs.readdir("hrady", (err, files) => {
  if (err) {
    console.error('Error reading directory:', err);
    return;
  }

  files.forEach((file) => {
    const filePath = path.join("hrady", file);
    fs.readFile(filePath, 'utf8', async (err, data) => {
      if (err) {
        console.error('Error reading file:', err);
        return;
      }

      try {
        const castleData = JSON.parse(data);
				print(data);
        // await insertCastleWithRelations(castleData);
        console.log(`Inserted castle from file: ${file}`);
      } catch (error) {
        console.error('Error inserting castle:', error);
      }
    });
  });
});
